cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java

```diff
@@ -0,0 +1,346 @@
+import io.fabric8.kubernetes.client.CustomResource;
+import io.fabric8.kubernetes.client.CustomResourceList;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.KubernetesClientException;
+import io.fabric8.kubernetes.client.Watcher;
+import io.fabric8.kubernetes.client.dsl.Resource;
+import io.fabric8.openshift.client.OpenShiftClient;
+import io.strimzi.api.kafka.KafkaConnectList;
+import io.strimzi.api.kafka.KafkaConnectS2IList;
+import io.strimzi.api.kafka.KafkaConnectorList;
+import io.strimzi.api.kafka.model.DoneableKafkaConnect;
+import io.strimzi.api.kafka.model.DoneableKafkaConnectS2I;
+import io.strimzi.api.kafka.model.DoneableKafkaConnector;
+import io.strimzi.api.kafka.model.KafkaConnect;
+import io.strimzi.api.kafka.model.KafkaConnectResources;
+import io.strimzi.api.kafka.model.KafkaConnectS2I;
+import io.strimzi.api.kafka.model.KafkaConnector;
+import io.strimzi.api.kafka.model.KafkaConnectorBuilder;
+import io.strimzi.api.kafka.model.KafkaConnectorSpec;
+import io.strimzi.api.kafka.model.status.KafkaConnectorStatus;
+import io.strimzi.api.kafka.model.status.Status;
+import io.strimzi.certs.CertManager;
+import io.strimzi.operator.PlatformFeaturesAvailability;
+import io.strimzi.operator.cluster.ClusterOperatorConfig;
+import io.strimzi.operator.cluster.model.ImagePullPolicy;
+import io.strimzi.operator.cluster.model.InvalidResourceException;
+import io.strimzi.operator.cluster.model.KafkaConnectCluster;
+import io.strimzi.operator.cluster.model.NoSuchResourceException;
+import io.strimzi.operator.cluster.model.StatusDiff;
+import io.strimzi.operator.cluster.operator.resource.ResourceOperatorSupplier;
+import io.strimzi.operator.common.AbstractOperator;
+import io.strimzi.operator.common.Annotations;
+import io.strimzi.operator.common.Reconciliation;
+import io.strimzi.operator.common.Util;
+import io.strimzi.operator.common.model.Labels;
+import io.strimzi.operator.common.operator.resource.ConfigMapOperator;
+import io.strimzi.operator.common.operator.resource.CrdOperator;
+import io.strimzi.operator.common.operator.resource.PodDisruptionBudgetOperator;
+import io.strimzi.operator.common.operator.resource.ServiceAccountOperator;
+import io.strimzi.operator.common.operator.resource.ServiceOperator;
+import io.strimzi.operator.common.operator.resource.StatusUtils;
+import io.vertx.core.CompositeFuture;
+import io.vertx.core.Future;
+import io.vertx.core.Vertx;
+import io.vertx.core.json.JsonObject;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public abstract class AbstractConnectOperator<C extends KubernetesClient, T extends CustomResource,
```

Comment:
@scholzj I had to completely rewrite this following your last review. Forgive me for feeling a little frustrated that you find the 2nd attempt complicated. 

Firstly it's worth bearing in mind that in your previous review you suggested that the `KafkaConnect` should take precedence over a same-named `KafkaConnectS2I` in owning `KafkaConnectors`. This complicates things because the operators are all stateless so when reconciling we need to check for existence of both connect and connects2i with a given name. Thus, in general, reconciliation actually involves three resources, and therefore there are 3 entry points to a reconciliation (connect, connects2i and connector watches/timers). 

The `AbstractConnectorOperator` is about abstracting over the duplicated code we had previously which was the same between the KafkaConnect operator and the KafkaConnectS2I operator. I did this because not creating such an abstraction would have meant a whole lot _more_ duplicated code. There's actually some scope to share more code, but I didn't attempt that in this PR.

I believe the main difference from the algorithm you describe is that there is no dedicated `KafkaConnectorOperator`. Instead there's a watch on connectors. The connector reconciliation logic is contained in the `reconcileConnectors()` method in `AbstractConnectorOperator`, which is shared between the KafkaConnect reconciliation and the reconciliation of a `KafkaConnector` triggered via a separate watch on connectors. I think it's this sharing of the code for reconciling a connector within a connect which is the difference between your suggestion and my implementation (your 2nd and 3rd bullets imply a duplication in making the REST API call).

A consequence of the approach I've taken here is that there's only one timed reconciliation going on, which reconciles connect and connectors. With your each connector would be reconciled on a timer and the connect on its own timer. 

But your comment made me realise that the connector watcher is needlessly triggering reconciliation of all connectors. I'll fix that so it only reconciles the one which was changed.

Reply:
Yeah, sorry if it frustrated you :-). And for the record I'm not saying you have to rewrite it as I say. I think this should work as well and since it already exists we should keep this.

Reply:
That's OK :-) I realise that the reasons why I implemented it like this aren't apparent just from reading the code. If it was done by someone else and I was reviewing it I'm sure I'd have similar concerns.